src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php

```diff
@@ -729,6 +729,103 @@ public function testAutowiring()
 
         $this->assertEquals('a', (string) $container->getDefinition('b')->getArgument(0));
     }
+
+    /**
+     * @dataProvider getSpecialCharServiceIds
+     */
+    public function testSpecialCharInServiceIdTriggersDeprecationInSet($id)
+    {
+        $container = new ContainerBuilder();
+
+        $deprecations = array();
+        set_error_handler(function ($type, $msg) use (&$deprecations) {
+            if (E_USER_DEPRECATED === $type) {
+                $deprecations[] = $msg;
+            }
+        });
+
+        $container->set($id, 'Foo');
+
+        restore_error_handler();
+
+        $this->assertCount(1, $deprecations);
+        $this->assertRegExp('/Using characters other than A-Z, 0-9, ., and _ in service ids \(".+" given\) is deprecated since Symfony 3.1 and will throw an exception in 4.0./', $deprecations[0]);
+    }
+
+    /**
+     * @dataProvider getSpecialCharServiceIds
+     */
+    public function testSpecialCharInServiceIdTriggersDeprecationInRegister($id)
+    {
+        $container = new ContainerBuilder();
+
+        $deprecations = array();
+        set_error_handler(function ($type, $msg) use (&$deprecations) {
+            if (E_USER_DEPRECATED === $type) {
+                $deprecations[] = $msg;
+            }
+        });
+
+        $container->register($id, 'foo');
+
+        restore_error_handler();
+
+        $this->assertCount(1, $deprecations);
+        $this->assertRegExp('/Using characters other than A-Z, 0-9, ., and _ in service ids \(".+" given\) is deprecated since Symfony 3.1 and will throw an exception in 4.0./', $deprecations[0]);
```

Comment:
I'd use a `AssertEquals` here rather than a regexp match (because we expect `foo` (or something else, see above ?) instead)